README.md
# Spidr

* http://spidr.rubyforge.org/
* http://github.com/postmodern/spidr
* http://github.com/postmodern/spidr/issues
* http://groups.google.com/group/spidr
* irc.freenode.net #spidr

## DESCRIPTION:

Spidr is a versatile Ruby web spidering library that can spider a site,
multiple domains, certain links or infinitely. Spidr is designed to be fast
and easy to use.

## FEATURES:

* Follows:
  * a tags.
  * iframe tags.
  * frame tags.
  * Cookie protected links.
  * HTTP 300, 301, 302, 303 and 307 Redirects.
  * HTTP Basic Auth protected links.
* Black-list or white-list URLs based upon:
  * URL scheme.
  * Host name
  * Port number
  * Full link
  * URL extension
* Provides call-backs for:
  * Every visited Page.
  * Every visited URL.
  * Every visited URL that matches a specified pattern.
  * Every URL that failed to be visited.
* Provides action methods to:
  * Pause spidering.
  * Skip processing of pages.
  * Skip processing of links.
* Restore the spidering queue and history from a previous session.
* Custom User-Agent strings.
* Custom proxy settings.
* HTTPS support.

## EXAMPLES:

Start spidering from a URL:

    Spidr.start_at('http://tenderlovemaking.com/')

Spider a host:

    Spidr.host('coderrr.wordpress.com')

Spider a site:

    Spidr.site('http://rubyflow.com/')

Spider multiple hosts:

    Spidr.start_at(
      'http://company.com/',
      :hosts => [
        'company.com',
        /host\d\.company\.com/
      ]
    )

Do not spider certain links:

    Spidr.site('http://matasano.com/', :ignore_links => [/log/])

Do not spider links on certain ports:

    Spidr.site(
      'http://sketchy.content.com/',
      :ignore_ports => [8000, 8010, 8080]
    )

Print out visited URLs:

    Spidr.site('http://rubyinside.org/') do |spider|
      spider.every_url { |url| puts url }
    end

Print out the URLs that could not be requested:

    Spidr.site('http://sketchy.content.com/') do |spider|
      spider.every_failed_url { |url| puts url }
    end

Search HTML and XML pages:

    Spidr.site('http://company.withablog.com/') do |spider|
      spider.every_page do |page|
        puts "[-] #{page.url}"

        page.search('//meta').each do |meta|
          name = (meta.attributes['name'] || meta.attributes['http-equiv'])
          value = meta.attributes['content']

          puts "    #{name} = #{value}"
        end
      end
    end

Print out the titles from every page:

    Spidr.site('http://www.rubypulse.com/') do |spider|
      spider.every_html_page do |page|
        puts page.title
      end
    end

Find what kinds of web servers a host is using, by accessing the headers:

    servers = Set[]

    Spidr.host('generic.company.com') do |spider|
      spider.all_headers do |headers|
        servers << headers['server']
      end
    end

Pause the spider on a forbidden page:

    spider = Spidr.host('overnight.startup.com') do |spider|
      spider.every_forbidden_page do |page|
        spider.pause!
      end
    end

Skip the processing of a page:

    Spidr.host('sketchy.content.com') do |spider|
      spider.every_missing_page do |page|
        spider.skip_page!
      end
    end

Skip the processing of links:

    Spidr.host('sketchy.content.com') do |spider|
      spider.every_url do |url|
        if url.path.split('/').find { |dir| dir.to_i > 1000 }
          spider.skip_link!
        end
      end
    end

## REQUIREMENTS:

* [nokogiri](http://nokogiri.rubyforge.org/) >= 1.2.0

## INSTALL:

    $ sudo gem install spidr

## LICENSE:

The MIT License

Copyright (c) 2008-2010 Hal Brodigan

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
### 0.2.2 / 2010-01-06

* Require Web Spider Obstacle Course (WSOC) >= 0.1.1.
* Integrated the new WSOC into the specs.
* Removed the built-in Web Spider Obstacle Course.
* Added {Spidr::Page#content_types}.
* Added {Spidr::Page#cookie}.
* Added {Spidr::Page#cookies}.
* Added {Spidr::Page#cookie_params}.
* Added {Spidr::Sanitizers}.
* Added {Spidr::SessionCache}.
* Added {Spidr::CookieJar} (thanks Nick Plante).
* Added {Spidr::AuthStore} (thanks Nick Plante).
* Added {Spidr::Agent#post_page} (thanks Nick Plante).
* Renamed Spidr::Agent#get_session to {Spidr::SessionCache#[]}.
* Renamed Spidr::Agent#kill_session to {Spidr::SessionCache#kill!}.

### 0.2.1 / 2009-11-25

* Added {Spidr::Events#every_ok_page}.
* Added {Spidr::Events#every_redirect_page}.
* Added {Spidr::Events#every_timedout_page}.
* Added {Spidr::Events#every_bad_request_page}.
* Added {Spidr::Events#every_unauthorized_page}.
* Added {Spidr::Events#every_forbidden_page}.
* Added {Spidr::Events#every_missing_page}.
* Added {Spidr::Events#every_internal_server_error_page}.
* Added {Spidr::Events#every_txt_page}.
* Added {Spidr::Events#every_html_page}.
* Added {Spidr::Events#every_xml_page}.
* Added {Spidr::Events#every_xsl_page}.
* Added {Spidr::Events#every_doc}.
* Added {Spidr::Events#every_html_doc}.
* Added {Spidr::Events#every_xml_doc}.
* Added {Spidr::Events#every_xsl_doc}.
* Added {Spidr::Events#every_rss_doc}.
* Added {Spidr::Events#every_atom_doc}.
* Added {Spidr::Events#every_javascript_page}.
* Added {Spidr::Events#every_css_page}.
* Added {Spidr::Events#every_rss_page}.
* Added {Spidr::Events#every_atom_page}.
* Added {Spidr::Events#every_ms_word_page}.
* Added {Spidr::Events#every_pdf_page}.
* Added {Spidr::Events#every_zip_page}.
* Fixed a bug where {Spidr::Agent#delay} was not being used to delay
  requesting pages.
* Spider `link` and `script` tags in HTML pages (thanks Nick Plante).

### 0.2.0 / 2009-10-10

* Added {URI.expand_path}.
* Added {Spidr::Page#search}.
* Added {Spidr::Page#at}.
* Added {Spidr::Page#title}.
* Added {Spidr::Agent#failures=}.
* Added a HTTP session cache to {Spidr::Agent}, per suggestion of falter.
  * Added Spidr::Agent#get_session.
  * Added Spidr::Agent#kill_session.
* Added {Spidr.proxy=}.
* Added {Spidr.disable_proxy!}.
* Aliased Spidr::Page#txt? to {Spidr::Page#plain_text?}.
* Aliased Spidr::Page#ok? to {Spidr::Page#is_ok?}.
* Aliased Spidr::Page#redirect? to {Spidr::Page#is_redirect?}.
* Aliased Spidr::Page#unauthorized? to {Spidr::Page#is_unauthorized?}.
* Aliased Spidr::Page#forbidden? to {Spidr::Page#is_forbidden?}.
* Aliased Spidr::Page#missing? to {Spidr::Page#is_missing?}.
* Split URL filtering code out of {Spidr::Agent} and into
  {Spidr::Filters}.
* Split URL / Page event code out of {Spidr::Agent} and into
  {Spidr::Events}.
* Split pause! / continue! / skip_link! / skip_page! methods out of
  {Spidr::Agent} and into {Spidr::Actions}.
* Fixed a bug in {Spidr::Page#code}, where it was not returning an Integer.
* Make sure {Spidr::Page#doc} returns Nokogiri::XML::Document objects for
  RSS/RDF/Atom pages as well.
* Fixed the handling of the Location header in {Spidr::Page#links}
  (thanks falter).
* Fixed a bug in {Spidr::Page#to_absolute} where trailing `/` characters on
  URI paths were not being preserved (thanks falter).
* Fixed a bug where the URI query was not being sent with the request
  in {Spidr::Agent#get_page} (thanks Damian Steer).
* Fixed a bug where SSL sessions were not being properly setup
  (thanks falter).
* Switched {Spidr::Agent#history} to be a Set, to improve search-time
  of the history (thanks falter).
* Switched {Spidr::Agent#failures} to a Set.
* Allow a block to be passed to {Spidr::Agent#run}, which will receive all
  pages visited.
* Allow Spidr::Agent#start_at and Spidr::Agent#continue! to pass blocks
  to {Spidr::Agent#run}.
* Made {Spidr::Agent#visit_page} public.
* Moved to YARD based documentation.

### 0.1.9 / 2009-06-13

* Upgraded to Hoe 2.0.0.
  * Use Hoe.spec instead of Hoe.new.
  * Use the Hoe signing task for signed gems.
* Added the Spidr::Agent#schemes and Spidr::Agent#schemes= methods.
* Added a warning message if 'net/https' cannot be loaded.
* Allow the list of acceptable URL schemes to be passed into
  {Spidr::Agent#initialize}.
* Allow history and queue information to be passed into
  {Spidr::Agent#initialize}.
* {Spidr::Agent#start_at} no longer clears the history or the queue.
* Fixed a bug in the sanitization of semi-escaped URLs.
* Fixed a bug where https URLs would be followed even if 'net/https'
  could not be loaded.
* Removed Spidr::Agent::SCHEMES.

### 0.1.8 / 2009-05-27

* Added the Spidr::Agent#pause! and Spidr::Agent#continue! methods.
* Added the Spidr::Agent#running? and Spidr::Agent#paused? methods.
* Added an alias for pending_urls to the queue methods.
* Added {Spidr::Agent#queue} to provide read access to the queue.
* Added {Spidr::Agent#queue=} and {Spidr::Agent#history=} for setting the
  queue and history.
* Added {Spidr::Agent#to_hash} which returns a Hash of the agents queue and
  history.
* Made {Spidr::Agent#enqueue} and {Spidr::Agent#queued?} public.
* Added more specs.

### 0.1.7 / 2009-04-24

* Added Spidr::Agent#all_headers.
* Fixed a bug where Page#headers was always `nil`.
* {Spidr::Spidr::Agent} will now follow the Location header in HTTP 300,
  301, 302, 303 and 307 Redirects.
* {Spidr::Agent} will now follow iframe and frame tags.

### 0.1.6 / 2009-04-14

* Added {Spidr::Agent#failures}, a list of URLs which could not be visited.
* Added {Spidr::Agent#failed?}.
* Added Spidr::Agent#every_failed_url.
* Added {Spidr::Agent#clear}, which clears the history and failures URL
  lists.
* Improved fault tolerance in {Spidr::Agent#get_page}.
  * If a Network or HTTP error is encountered, the URL will be added to
    the failures list and the next URL will be visited.
* Fixed a typo in Spidr::Agent#ignore_exts_like.
* Updated the Web Spider Obstacle Course with links that always fail to be

lib/spidr/agent.rb
require 'spidr/sanitizers'
require 'spidr/filters'
require 'spidr/events'
require 'spidr/actions'
require 'spidr/page'
require 'spidr/session_cache'
require 'spidr/cookie_jar'
require 'spidr/auth_store'
require 'spidr/spidr'

require 'net/http'
require 'set'

module Spidr
  class Agent

    include Sanitizers
    include Filters
    include Events
    include Actions

    # HTTP Host Header to use
    attr_accessor :host_header

    # HTTP Host Headers to use for specific hosts
    attr_reader :host_headers

    # User-Agent to use
    attr_accessor :user_agent

    # HTTP Authentication credentials
    attr_accessor :authorized

    # Referer to use
    attr_accessor :referer

    # Delay in between fetching pages
    attr_accessor :delay

    # History containing visited URLs
    attr_reader :history

    # List of unreachable URLs
    attr_reader :failures

    # Queue of URLs to visit
    attr_reader :queue

    # Cached cookies
    attr_reader :cookies

    #
    # Creates a new Agent object.
    #
    # @param [Hash] options
    #   Additional options
    #
    # @option options [Hash] :proxy (Spidr.proxy)
    #   The proxy information to use.
    #
    # @option :proxy [String] :host
    #   The host the proxy is running on.
    #
    # @option :proxy [Integer] :port
    #   The port the proxy is running on.
    #
    # @option :proxy [String] :user
    #   The user to authenticate as with the proxy.
    #
    # @option :proxy [String] :password
    #   The password to authenticate with.
    #
    # @option options [String] :host_header
    #   The HTTP Host header to use with each request.
    #
    # @option options [Hash{String,Regexp => String}] :host_headers
    #   The HTTP Host headers to use for specific hosts.
    #
    # @option options [String] :user_agent (Spidr.user_agent)
    #   The User-Agent string to send with each requests.
    #
    # @option options [String] :referer
    #   The Referer URL to send with each request.
    #
    # @option options [Integer] :delay (0)
    #   The number of seconds to pause between each request.
    #
    # @option options [Set, Array] :queue
    #   The initial queue of URLs to visit.
    #
    # @option options [Set, Array] :history
    #   The initial list of visited URLs.
    #
    # @yield [agent]
    #   If a block is given, it will be passed the newly created agent
    #   for further configuration.
    #
    # @yieldparam [Agent] agent
    #   The newly created agent.
    #
    def initialize(options={},&block)
      @host_header = options[:host_header]
      @host_headers = {}

      if options[:host_headers]
        @host_headers.merge!(options[:host_headers])
      end

      @user_agent = (options[:user_agent] || Spidr.user_agent)
      @referer = options[:referer]

      @sessions = SessionCache.new(options[:proxy] || Spidr.proxy)
      @cookies = CookieJar.new
      @authorized = AuthStore.new

      @running = false
      @delay = (options[:delay] || 0)
      @history = Set[]
      @failures = Set[]
      @queue = []

      super(options)

      block.call(self) if block
    end

    #
    # Creates a new agent and begin spidering at the given URL.
    #
    # @param [URI::HTTP, String] url
    #   The URL to start spidering at.
    #
    # @param [Hash] options
    #   Additional options. See {Agent#initialize}.
    #
    # @yield [agent]
    #   If a block is given, it will be passed the newly created agent
    #   before it begins spidering.
    #
    # @yieldparam [Agent] agent
    #   The newly created agent.
    #
    def self.start_at(url,options={},&block)
      self.new(options) do |spider|
        block.call(spider) if block

        spider.start_at(url)
      end
    end

    #
    # Creates a new agent and spiders the given host.
    #
    # @param [String]
    #   The host-name to spider.
    #
    # @param [Hash] options
    #   Additional options. See {Agent#initialize}.
    #
    # @yield [agent]
    #   If a block is given, it will be passed the newly created agent
    #   before it begins spidering.
    #
    # @yieldparam [Agent] agent
    #   The newly created agent.
    #
    def self.host(name,options={},&block)
      self.new(options.merge(:host => name)) do |spider|
        block.call(spider) if block

        spider.start_at("http://#{name}/")
      end
    end

    #
    # Creates a new agent and spiders the web-site located at the given URL.
    #
    # @param [URI::HTTP, String] url
    #   The web-site to spider.
    #
    # @param [Hash] options
    #   Additional options. See {Agent#initialize}.
    #
    # @yield [agent]
    #   If a block is given, it will be passed the newly created agent
    #   before it begins spidering.
    #
    # @yieldparam [Agent] agent
    #   The newly created agent.
    #
    def self.site(url,options={},&block)
      url = URI(url.to_s)

      return self.new(options.merge(:host => url.host)) do |spider|
        block.call(spider) if block

        spider.start_at(url)
      end
    end

    #
    # Clears the history of the agent.
    #
    def clear
      @queue.clear
      @history.clear
      @failures.clear
      return self
    end

    #
    # Start spidering at a given URL.
    #
    # @param [URI::HTTP, String] url
    #   The URL to start spidering at.
    #
    # @yield [page]
    #   If a block is given, it will be passed every page visited.
    #
    # @yieldparam [Page] page
    #   A page which has been visited.
    #
    def start_at(url,&block)
      enqueue(url)

      return run(&block)
    end

    #
    # Start spidering until the queue becomes empty or the agent is
    # paused.
    #
    # @yield [page]
    #   If a block is given, it will be passed every page visited.
    #
    # @yieldparam [Page] page
    #   A page which has been visited.
    #
    def run(&block)
      @running = true

      until (@queue.empty? || paused?)
        begin
          visit_page(dequeue,&block)
        rescue Actions::Paused
          return self
        rescue Actions::Action
        end
      end

      @running = false

      @sessions.clear
      return self
    end

    #
    # Determines if the agent is running.
    #
    # @return [Boolean]
    #   Specifies whether the agent is running or stopped.
    #
    def running?
      @running == true
    end

    #
    # The proxy information the agent uses.
    #
    # @return [Hash]
    #   The proxy information.
    #
    # @see SessionCache#proxy
    #
    # @since 0.2.2
    #
    def proxy
      @sessions.proxy
    end

    #
    # Sets the proxy information that the agent uses.
    #
    # @param [Hash] new_proxy
    #   The new proxy information.
    #
    # @return [Hash]
    #   The new proxy information.
    #
    # @see SessionCache#proxy=
    #
    # @since 0.2.2
    #
    def proxy=(new_proxy)
      @sessions.proxy = new_proxy
    end

    #
    # Sets the history of URLs that were previously visited.
    #
    # @param [#each] new_history
    #   A list of URLs to populate the history with.
    #
    # @return [Set<URI::HTTP>]
    #   The history of the agent.
    #
    # @example
    #   agent.history = ['http://tenderlovemaking.com/2009/05/06/ann-nokogiri-130rc1-has-been-released/']
    #
    def history=(new_history)
      @history.clear

      new_history.each do |url|
        @history << unless url.kind_of?(URI)
                      URI(url.to_s)
                    else
                      url
                    end
      end

      return @history
    end

    alias visited_urls history

    #
    # Specifies the links which have been visited.
    #
    # @return [Array<String>]
    #   The links which have been visited.
    #
    def visited_links
      @history.map { |url| url.to_s }
    end

    #
    # Specifies all hosts that were visited.
    #
    # @return [Array<String>]
    #   The hosts which have been visited.
    #
    def visited_hosts
      visited_urls.map { |uri| uri.host }.uniq
    end

    #
    # Determines whether a URL was visited or not.
    #
    # @param [URI::HTTP, String] url
    #   The URL to search for.
    #
    # @return [Boolean]
    #   Specifies whether a URL was visited.
    #
    def visited?(url)
      url = URI(url.to_s) unless url.kind_of?(URI)

      return @history.include?(url)
    end

    #
    # Sets the list of failed URLs.
    #
    # @param [#each]
    #   The new list of failed URLs.
    #
    # @return [Array<URI::HTTP>]
    #   The list of failed URLs.
    #
    # @example
    #   agent.failures = ['http://localhost/']
    #
    def failures=(new_failures)
      @failures.clear

      new_failures.each do |url|
        @failures << unless url.kind_of?(URI)
                    URI(url.to_s)
                  else
                    url
                  end
      end

      return @failures
    end

    #
    # Determines whether a given URL could not be visited.
    #
    # @param [URI::HTTP, String] url
    #   The URL to check for failures.
    #
    # @return [Boolean]
    #   Specifies whether the given URL was unable to be visited.
    #
    def failed?(url)
      url = URI(url.to_s) unless url.kind_of?(URI)

      return @failures.include?(url)
    end

    alias pending_urls queue

    #
    # Sets the queue of URLs to visit.
    #
    # @param [#each]
    #   The new list of URLs to visit.
    #
    # @return [Array<URI::HTTP>]
    #   The list of URLs to visit.
    #
    # @example
    #   agent.queue = ['http://www.vimeo.com/', 'http://www.reddit.com/']
    #
    def queue=(new_queue)
      @queue.clear

      new_queue.each do |url|
        @queue << unless url.kind_of?(URI)
                    URI(url.to_s)
                  else
                    url
                  end
      end

      return @queue
    end

    #
    # Determines whether a given URL has been enqueued.
    #
    # @param [URI::HTTP] url
    #   The URL to search for in the queue.
    #
    # @return [Boolean]
    #   Specifies whether the given URL has been queued for visiting.
    #
    def queued?(url)
      @queue.include?(url)
    end

    #
    # Enqueues a given URL for visiting, only if it passes all of the
    # agent's rules for visiting a given URL.
    #
    # @param [URI::HTTP, String] url
    #   The URL to enqueue for visiting.
    #
    # @return [Boolean]
    #   Specifies whether the URL was enqueued, or ignored.
    #
    def enqueue(url)
      url = sanitize_url(url)

      if (!(queued?(url)) && visit?(url))
        link = url.to_s

        begin
          @every_url_blocks.each { |block| block.call(url) }

          @urls_like_blocks.each do |pattern,blocks|
            if ((pattern.kind_of?(Regexp) && link =~ pattern) || pattern == link || pattern == url)
              blocks.each { |url_block| url_block.call(url) }
            end
          end
        rescue Actions::Paused => action
          raise(action)
        rescue Actions::SkipLink
          return false
        rescue Actions::Action
        end

        @queue << url
        return true
      end

      return false
    end

    #
    # Requests and creates a new Page object from a given URL.
    #
    # @param [URI::HTTP] url
    #   The URL to request.
    #
    # @yield [page]
    #   If a block is given, it will be passed the page that represents the
    #   response.
    #
    # @yieldparam [Page] page
    #   The page for the response.
    #
    # @return [Page, nil]
    #   The page for the response, or `nil` if the request failed.
    #
    def get_page(url,&block)
      url = URI(url.to_s)

      prepare_request(url) do |session,path,headers|
        new_page = Page.new(url,session.get(path,headers))

        # save any new cookies
        @cookies.from_page(new_page)

        block.call(new_page) if block
        return new_page
      end
    end

    #
    # Posts supplied form data and creates a new Page object from a given URL.
    #
    # @param [URI::HTTP] url
    #   The URL to request.
    #
    # @param [String] post_data
    #   Form option data.
    #
    # @yield [page]
    #   If a block is given, it will be passed the page that represents the
    #   response.
    #
    # @yieldparam [Page] page
    #   The page for the response.
    #
    # @return [Page, nil]
    #   The page for the response, or `nil` if the request failed.
    #
    # @since 0.2.2
    #
    def post_page(url,post_data='',&block)
      url = URI(url.to_s)

      prepare_request(url) do |session,path,headers|
        new_page = Page.new(url,session.post(path,post_data,headers))

        # save any new cookies
        @cookies.from_page(new_page)

        block.call(new_page) if block
        return new_page
      end
    end

    #
    # Visits a given URL, and enqueus the links recovered from the URL
    # to be visited later.
    #
    # @param [URI::HTTP, String] url
    #   The URL to visit.
    #
    # @yield [page]
    #   If a block is given, it will be passed the page which was visited.
    #
    # @yieldparam [Page] page
    #   The page which was visited.
    #
    # @return [Page, nil]
    #   The page that was visited. If `nil` is returned, either the request
    #   for the page failed, or the page was skipped.
    #
    def visit_page(url,&block)
      url = URI(url.to_s) unless url.kind_of?(URI)

      get_page(url) do |page|
        @history << page.url

        begin
          @every_page_blocks.each { |page_block| page_block.call(page) }

          block.call(page) if block
        rescue Actions::Paused => action
          raise(action)
        rescue Actions::SkipPage
          return nil
        rescue Actions::Action
        end

        page.urls.each { |next_url| enqueue(next_url) }
      end
    end

    #
    # Converts the agent into a Hash.
    #
    # @return [Hash]
    #   The agent represented as a Hash containing the `history` and
    #   the `queue` of the agent.
    #
    def to_hash
      {:history => @history, :queue => @queue}
    end

    protected

    #
    # Normalizes the request path and grabs a session to handle page
    # get and post requests.
    #
    # @param [URI::HTTP] url
    #   The URL to request.
    #
    # @yield [request]
    #   A block whose purpose is to make a page request.
    #
    # @yieldparam [Net::HTTP] session
    #   An HTTP session object.
    #
    # @yieldparam [String] path
    #   Normalized URL string.
    #
    # @yieldparam [Hash] headers
    #   A Hash of request header options.
    #
    # @since 0.2.2
    #
    def prepare_request(url,&block)
      host = url.host
      port = url.port

      unless url.path.empty?
        path = url.path
      else
        path = '/'
      end

      # append the URL query to the path
      path += "?#{url.query}" if url.query

      # set any additional HTTP headers
      headers = {}

      unless @host_headers.empty?
        @host_headers.each do |name,header|
          if host.match(name)
            headers['Host'] = header
            break
          end
        end
      end

      headers['Host'] ||= @host_header if @host_header
      headers['User-Agent'] = @user_agent if @user_agent
      headers['Referer'] = @referer if @referer

      if (authorization = @authorized.for_url(url))
        headers['Authorization'] = "Basic #{authorization}"
      end

      if (header_cookies = @cookies.for_host(url.host))
        headers['Cookie'] = header_cookies
      end

      begin
        sleep(@delay) if @delay > 0

        block.call(@sessions[url],path,headers)
      rescue SystemCallError,
             Timeout::Error,
             SocketError,
             Net::HTTPBadResponse,
             IOError

        @sessions.kill!(url)

        failed(url)
        return nil
      end
    end

    #
    # Dequeues a URL that will later be visited.
    #
    # @return [URI::HTTP]
    #   The URL that was at the front of the queue.
    #
    def dequeue
      @queue.shift
    end

    #
    # Determines if a given URL should be visited.
    #
    # @param [URI::HTTP] url
    #   The URL in question.
    #
    # @return [Boolean]
    #   Specifies whether the given URL should be visited.
    #
    def visit?(url)
      (!(visited?(url)) &&
       visit_scheme?(url.scheme) &&
       visit_host?(url.host) &&
       visit_port?(url.port) &&
       visit_link?(url.to_s) &&
       visit_ext?(url.path))
    end

    #
    # Adds a given URL to the failures list.
    #
    # @param [URI::HTTP] url
    #   The URL to add to the failures list.
    #
    def failed(url)
      @failures << url
      @every_failed_url_blocks.each { |block| block.call(url) }
      return true
    end

  end
end
